### ISSUE TITLE
Extra page in table of contents. (<tocpagebreak>)

### ISSUE THREAD

**Issue by tomacadi** [mpdf/mpdf#227]
### I have this problem/would like to have this functionality
The table of contents has 2 pages by default. In a short document, this is not necessary.
Is there a way to suppress this second page?

### This is a HTML code snippet I use

```
<body>
<tocpagebreak></tocpagebreak>
<page>
<tocentry content="toc name" level="0"></tocentry>
toc name
</page>
</body>
```

**Comment by tomacadi** [comment]
[extra_page.pdf](https://github.com/mpdf/mpdf/files/456576/extra_page.pdf)
Added PDF generated by this snippet.

**Comment by bernhardh** [comment]
I have the same problem.  Short document, but one extra (blank page)

**Comment by bigattichouse** [comment]
If you have ANY content after the tocpagebreak (like and </dev> or a space) - it gets pushed to the next page... I also had <pagebreak> tags at the beginning of each chapter that were adding anextranesou page.

Now to get the TOC paging to not be a huge mess.

**Comment by nsmith-sas** [comment]
Yeah this is much needed. It creates two pages after the TOC page. A second page for the TOC which isn't needed at all. Then it starts a new page for new content. I have custom headers on each page so the entire TOC function is broken for what I want. All I need is a one page TOC

**Comment by bigattichouse** [comment]
Remember, there can be NOTHING after the tag if you want to prevent the blank. so a tocpagebreak, and then your next page immediately after.

Try setting the # of columns on your ToC to get more columns, and try changing the mpdf.css to make the text smaller.. you can probably get it down to 1 page.

**Comment by bigattichouse** [comment]
My tocpagebreak:
I had to manually change the start and end pages to get it to work (I don't know how you'd get it to work properly in an automated fashion)
<tocpagebreak toc-preHTML="&lt;H2&gt;Table of Contents&lt;/H2&gt;" toc-postHTML="" links="ON" toc-sheet-size="Letter" toc-resetpagenum="2" resetpagenum="6" toc-bookmarkText="Table of Contents"/>

**Comment by nsmith-sas** [comment]
If you just call the tocpagebreak with bare minimum code. the TOC is on page 1, and page 2, then it starts a new page. The way my code is, I set a header, then start a new page, then set a footer in a for loop. However, with the TOC function creating a new page makes it so I can't define a header before a page creation. Each page has a dynamic header. So if there is a way to define a header after a page creation then it will just be a pointless second page for the TOC with no content on it.

**Comment by jakejackson1** [comment]
@nsmith-sas You define the header using the `tocpagebreak` tag, as oppose to setting it before hand. Refer to the docs for the correct syntax: http://mpdf.github.io/reference/html-control-tags/tocpagebreak.html

**Comment by Nick74k** [comment]
The same problem with mPDF 7.0
```
<tocentry content="P1" />Page 1
<pagebreak /><tocentry content="P2" />Page 2
<pagebreak /><tocpagebreak links="1" />
```
This code gives P1,P2,TOC and extra blank page.
```
<tocentry content="P1" />Page 1
<pagebreak /><tocentry content="P2" />Page 2
<tocpagebreak links="1" />
```
This gives P1,TOC,P2.
Please help how to get P1,P2,TOC???

**Comment by carstengrimm** [comment]
depending on content i have one or more empty pages after TOC while not using html code but the php equivalent....

**Comment by JonGerlach** [comment]
I was having the same problem; an extra, blank page was added after my table of contents. I was able to solve this by deleting the blank page just before outputting the PDF. This is definitely a workaround not a solution, but it's getting the job done for what I need.

Set `$toc_placeholder_page ` to the current page in the document, then add a TOC.
<pre>
$toc_placeholder_page = $mpdf->page + 1;
$mpdf->TOCpagebreak();
</pre>

Delete the extra page, then output the document.
<pre>
$mpdf->DeletePages($toc_placeholder_page);
$mpdf->Output();
</pre>

This is working for me regardless of where the TOC falls in the document or if if it's a single or multiple page TOC. It also corrects the page numbering issues I was running into.

**Comment by sijovijayan** [comment]
==== Try this ======
https://stackoverflow.com/questions/31025178/how-to-delete-blank-page-in-mpdf/45935484#45935484

**Comment by damelco** [comment]
this was fixed?
in what version?

**Comment by Sunchock** [comment]
I have the same problem
`$mpdf = new \Mpdf\Mpdf();

$html = '
<html>
<head>
<style>
	@page blank {
		size: auto;
		header: _blank;
		footer: _blank;
	}
    @page chapter1 {
    	header: html_myHeader1;
		footer: html_myFooter1;
    }
    @page chapter2 {
        header: html_myHeader2;
		footer: html_myFooter2;
    }
    div.blank {
    	page-break-before: always;
    	page: blank;
    }
    div.chapter1 {
        /* page-break-before: always; */
        page: chapter1;
    }
    div.chapter2 {
        /*page-break-before: always;*/
        page: chapter2;
    }
</style>
</head>

<body>

    <htmlpageheader name="myHeader1" style="display:none">
        <div style="text-align: right; border-bottom: 1px solid #000000; font-weight: bold; font-size: 10pt;">
            myHeader1
        </div>
    </htmlpageheader>

    <htmlpageheader name="myHeader2" style="display:none">
        <div style="border-bottom: 1px solid #000000; font-weight: bold;  font-size: 10pt;">
            myHeader2
        </div>
    </htmlpageheader>

    <htmlpagefooter name="myFooter1" style="display:none">
        <table width="100%">
            <tr>
                <td width="33%">
                    <span style="font-weight: bold; font-style: italic;">{DATE j-m-Y}</span>
                </td>
                <td width="33%" align="center" style="font-weight: bold; font-style: italic;">
                    {PAGENO}/{nbpg}
                </td>
                <td width="33%" style="text-align: right;">
                    myFooter1
                </td>
            </tr>
        </table>
    </htmlpagefooter>

    <htmlpagefooter name="myFooter2" style="display:none">
        <table width="100%">
            <tr>
                <td width="33%">myFooter2</td>
                <td width="33%" align="center">{PAGENO}/{nbpg}</td>
                <td width="33%" style="text-align: right;">{DATE j-m-Y}</td>
            </tr>
        </table>
    </htmlpagefooter>

	<div class="blank">Cover page</div>
    
    <tocpagebreak toc-page-selector="blank" />
    
    <div class="chapter1">Text<tocentry content="Chapter 1"> of Chapter 1</div>
    
    <div class="chapter2">Text<tocentry content="Chapter 2"> of Chapter 2</div>
    
    <div class="chapter1">Text<tocentry content="Chapter 3"> of Chapter 3</div>
</body>
</html>';

$mpdf->WriteHTML($html);
$mpdf->Output();`

This code gives Cover, Toc, EXTRA BLANK PAGE, p1,p2,p3.

**Comment by Sunchock** [comment]
> this was fixed?
> in what version?

This problem is still not resolved in May 2020

**Comment by Klap-in** [comment]
Contributing is possible. It depends mainly on volunteers here. Thanks.

**Comment by tobiashammarstrand** [comment]
@Sunchock, have you find any solution yet to your problem with blank pages? I have same html-structure like you and have same problem. New page before next section after TOC.

**Comment by Sunchock** [comment]
> @Sunchock, have you find any solution yet to your problem with blank pages? I have same html-structure like you and have same problem. New page before next section after TOC.

The only solution I found was to delete the extra blank page `$mpdf->DeletePages(2);` but if your table of content is bigger than one page, the problem may be coming again.
So I choose to let a blank page between toc and pages to be compatible with big toc.

**Comment by tobiashammarstrand** [comment]
@Sunchock , ok. I tried to figure out how the blank pages are inserted. And in my case, it is the next CSS-Page that starts with a blank page.  I did a hack and added an if statement in the BlockTag.php file to prevent extra page if the next CSS-Page name is equal to my first section after the TOC.

Starting from line 320 in BlockTag.php
```
 /* -- CSS-PAGE -- */
// Must Add new page if changed page properties
elseif (!isset($this->mpdf->page_box['current']) || $pagesel != $this->mpdf->page_box['current'] ) {
 if($pagesel != "section_content"){
$this->mpdf->AddPage($this->mpdf->CurOrientation, '', '', '', '', '', '', '', '', '', '', '', '', '', '', 0, 0, 0, 0, $pagesel);
 }
}
/* -- END CSS-PAGE -- */
```

**Comment by Sunchock** [comment]
@tobiashammarstrand 
Thank you for sharing. I used mpdf for my company and I moved on another project so I can't help you anymore
I hope this thread can help future mpdf users !

**Comment by Fulcanellii** [comment]
I had the same problem

CSS
`
 .new-page {
            page-break-before: always;
        }
`

HTML 

```

<tocpagebreak links="1" toc-preHTML="&lt;h2&gt;Table of Contents&lt;/h1&gt;" toc-margin-top="155px" />

<section class="new-page">

    <tocentry content="Definitions & Abbreviations" />

    <h3 class="page__body__title">Definitions & Abbreviations</h3>

</section>

<section class="new-page">

    <tocentry content="1. Ship particulars" />

    <h3 class="page__body__title">1. Ship particulars</h3>

</section>
```

Decision
Remove "new-page" class for the first section after <tocpagebreak />
